sets.c

```diff
@@ -0,0 +1,222 @@
+ */
+static int _size_t_cmp(const void *_i1, const void *_i2)
+{
+    const size_t i1 = *(const size_t *)_i1, i2 = *(const size_t *)_i2;
+
+    if      (i1 <  i2) return -1;
+    else if (i1 == i2) return  0;
+    else               return  1;
+}
+
+struct set *parse_uintset(const char *lptr, size_t maxsz)
+{
+    const char *tok, *tmpnum, *tmpnum_end;
+    bool        got_num;
+    size_t      last_num, arr_szfilled, arr_maxsz;
+    size_t     *valarr;
+    size_t      vaidx = 0;
+    struct set *ret_set;
+    uservalue_t tokval;
+
+    assert(lptr);
+
+    enum {
+        NIL = -1,
+        NUM_TOK,
+        RNG_TOK,
+        CMA_TOK,
+    } last_type = NIL;
+
+#if __SIZEOF_SIZE_T__ == 8
+#define get_val(val) ((val).uint64_value)
+#elif __SIZEOF_SIZE_T__ == 4
+#define get_val(val) ((val).uint32_value)
+#elif __SIZEOF_SIZE_T__ == 2
+#define get_val(val) ((val).uint16_value)
+#elif __SIZEOF_SIZE_T__ == 1
+#define get_val(val) ((val).uint8_value)
+#else
+#error unsupported word size
+#endif
+
+    /* allocate space for value pointers */
+    if ((valarr = calloc(DEFAULT_UINTLS_SZ, sizeof(size_t))) == NULL)
+        return NULL;
+    if ((ret_set = calloc(1, sizeof(struct set))) == NULL) {
+        free(valarr);
+        return NULL;
+    }
+
+    arr_szfilled = 0;
+    arr_maxsz = DEFAULT_UINTLS_SZ;
+    got_num = false;
+
+/* NOTE: used locally. fragile, make sure to not use it with anything fancy. */
+#define INC_ARRAY_SZ                                                                 \
+    do {                                                                             \
+        if (arr_maxsz > maxsz / 2 ||                                          \
+            (valarr = realloc(valarr, (arr_maxsz *= 2) * sizeof(size_t))) == NULL) \
+            goto error;                                                              \
+    } while (0)
+
+    for (tok = lptr; *tok; tok++) {
+        /* skip spaces */
+        while (tok && isspace(*tok))
+            tok++;
+
+        /* check if we just went over trailing space */
+        if (!tok || !*tok)
+            break;
+
+        assert(arr_szfilled <= arr_maxsz); // should never fail
+        if (arr_szfilled == arr_maxsz)
+            INC_ARRAY_SZ;
+
+        switch (*tok) {
+        case ',':
+            if (last_type == RNG_TOK)
+                goto error;
+            last_type = CMA_TOK;
+            continue;
+        case '.':
+            if (last_type == CMA_TOK || last_type == RNG_TOK)
+                goto error;
+            /* check for consecutive `..` */
+            if (tok != lptr && tok[-1] == '.')
+                last_type = RNG_TOK;
+            continue;
+        }
+
+        if (isdigit(*tok)) {
+            /* find the end of this number in string */
+            tmpnum_end = tok;
+            for ( ; isdigit(*tmpnum_end); tmpnum_end++)
+                ;
+            tmpnum = strndup(tok, tmpnum_end - tok);
```

Comment:
IMHO, this leaks `tmpnum`, as it lack a `free()`.

Reply:
Fixed.

Reply:
Yeah, looks good to me too.
I'll pass this whole thing under valgrind anyway when we're done tweaking.